Basic secure array for areg

dyfet opened the pull request:
Initial pass at a secure_array areg sty;le...

dyfet committed:
Basic secure array for areg

dyfet committed:
Embedded use case for loading raw keys, like from firmware

dyfet committed:
Improved assignment operator

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,175 @@
+    TESecureArray() noexcept = default;
+    TESecureArray(const TESecureArray& from) noexcept : _empty(from._empty) {
+        if (!_empty)
+            memcpy(_data, from._data, SIZE);
```
instead `memcpy` would be nice to use [`memcpy_s`](https://en.cppreference.com/w/c/string/byte/memcpy.html) or [`NEMemory::memCopy`](https://github.com/dyfet/areg-sdk/blob/master/framework/areg/base/NEMemory.hpp#L761), which in principle is the same as `memcpy_s`.

dyfet replied:
memcpy_s is not cross platform....there is a memcpy in minicrypt, but it is not built if AREG_CRYPTO=0. But we could use it if we always require AREG_CRYPTO to use all of ccrypto including headers.

aregtech replied:
> memcpy_s is not cross platform

Oh, interesting. [This cppreference.com](https://en.cppreference.com/w/c/string/byte/memcpy.html) doc does not say anything about this. It says that `since C11`. Do you have source about this?

dyfet replied:
"...As with all bounds-checked functions, memcpy_s is only guaranteed to be available if __STDC_LIB_EXT1__ is defined by the implementation and if the user defines __STDC_WANT_LIB_EXT1__ to the integer constant 1 before including [<string.h>](https://en.cppreference.com/w/c/header/string.html)..."

It "may" be present, but it also may not. I ran into a similar problem with C11 threads where mingw posix did NOT support the C11 thread standard, not even the latest releases of gcc, and the BSD libc also did NOT support it, which meant that libcpr (a pure c library) had to use pthread to be portable. That is why I am more caution using annex / extension things now.

aregtech replied:
ah, I should be more attentive when reading docs :D 
In this case, I would suggest to use the `NEMemory::memCopy` method and then there decide whether it uses `memcpy` or `memcpy_s`.
I'm saying this, because use of `memcpy` displays warning message suggesting to use additional compiler flag to escape warning, which I don't want.

dyfet replied:
Indeed, this is often why I do my own wrappers for memory operations; it's easier to fix it in a common wrapper and fix that more later if needed than having to adjust each call site.

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    // Load a key value from a physical address object
+    TESecureArray(const std::byte *from) noexcept : _empty(from == nullptr ? true : false) {
```
`_empty(from == nullptr ? true : false)` is same as `_empty(from == nullptr)`

dyfet replied:
True... and that could be true for the related constructor, too...

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,175 @@
+        if (len) {
+            memcpy(&_data, from.data(), len);
+            // forces erasure of origin data...
+            auto wp = const_cast<void *>(reinterpret_cast<const void *>(from.data()));
```
In this case, do we really need this declaration `explicit TESecureArray(const BINARY& from) noexcept`, which not declaring without `const` --> `explicit TESecureArray(BINARY& from) noexcept`?

dyfet replied:
I was likely avoiding automatic conversions, but maybe that is entirely okay. for this.

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,175 @@
+
+    // memory safe copy so we can remove [] operators
+    template <typename BINARY>
+    auto merge(std::size_t offset, const BINARY& from) -> std::size_t {
```
is this really merging or copying? it seems that the function copies data starting from very first element.
or if merging, then `memcpy_s(_data + offset, SIZE - offset, from.data(), count);`

dyfet replied:
Thats a good question...

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    auto operator=(const std::byte *from) noexcept -> auto& {
+        _empty = from == nullptr ? true : false;
```
`_empty = from == nullptr;`

aregtech reviewed (changes requested):
Thank you very much.
Please see comments.

dyfet committed:
Resp;ve nullptr empty

dyfet committed:
Use minicrypto memori wrappers

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+    // Load a key value from a physical address object
+    TESecureArray(const std::byte *from) noexcept : _empty(from == nullptr) {
+        if (from != nullptr)
+            MiniCrypt::memcpy(_data, from, SIZE);
```
Can it be possible that the length of the `from` is not equal to `SIZE`? Note, that here if the length of `from` is less than `SIZE`, we'll copy garbage.

dyfet replied:
Yea, it was not a good idea for me to add this...it should be gone now

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+    explicit TESecureArray(const BINARY& from) noexcept {
+        auto len = std::min(SIZE, from.size());
+        if (len) {
+            MiniCrypt::memcpy(&_data, from.data(), len);
```
Not sure that `&_data` is correct. In principle, you set the address of the pointer `_data`?
Please re-check, I think the right call is `MiniCrypt::memcpy(_data, from.data(), len);`
Or maybe you planned `&_data[0]`?

dyfet replied:
We now have minicrypt helpers used. I think _data resolves to a pointer correctly....

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+            MiniCrypt::memcpy(&_data, from.data(), len);
+            // forces erasure of origin data...
+            auto wp = const_cast<void *>(reinterpret_cast<const void *>(from.data()));
+            MiniCrypt::memset(wp, 0, len);
```
Please note, that if `from` has bigger size, not all data will be set `0`, because of this: `auto len = std::min(SIZE, from.size());`
In other words, if `from` has size `10` and `this` has size `5`, only first 5 entries will set `0`. Please re-check whether this is what have planned.

dyfet replied:
Indeed, I have to think about this. I was not sure I even wanted to do a forced update on a const object to begin with ;). And we can't "clear" it, only erase bytes. So maybe this should not try writing back at all...

aregtech replied:
I don't know what kind of plans you have for this object. I just note what seems is not clear :) 
Pointing to re-check.

dyfet replied:
More to the point, I really use this a lot in my code base as a template type adapter for binary containers that I might pipeline. This is a pattern not used in areg, and there is no existing code to support, so I removed it and it's associated operator for now. We can always add stuff back later if we find we actually need it. The existing copy method provides a safe mechanism to load other data, which makes the unsafe pure byte ptr loader also unnecessary, but that is already gone.

I think this is closer to what we want, but yes, I think I should also add the comments from your new pr, or we can adjust your pr dropping these things there as well. I am open to either option.

aregtech replied:
> We can always add stuff back later if we find we actually need it.  

Absolutely, I completely agree.  

In general, I don’t mind if classes expose more methods than currently required by Areg’s logic, especially when it’s a template or when extra utilities could be helpful for developers.  
The key point is just to keep the interface clear and intuitive so that developers aren’t left guessing


> I think this is closer to what we want, but yes, I think I should also add the comments from your new PR, or we can adjust your PR dropping these things there as well. I am open to either option.  

Since you’re leading development on this module, I’d really prefer to follow your direction.  
If you’re comfortable with the suggestions I made in the other PR, I can take care of adapting everything to match Areg’s style and structure.  

Just let me know once you consider this version feature-complete or stable, and I’ll handle the cosmetic and structural adjustments. The main thing for me is that you’re comfortable with the approach and how it fits into the overall framework. The rest is easy :)

dyfet replied:
I feel this version is indeed now feature complete now, and stable. I had no further changes contemplated other than what might be found when adding other things later.

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+    auto operator=(const std::byte *from) noexcept -> auto& {
+        _empty = (from == nullptr);
+        if (from != nullptr)
+            MiniCrypt::memcpy(_data, from, SIZE);
```
Same is here. Are we sure that the length of `from` is not less than `SIZE`?

dyfet replied:
I think this is gone now...

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+
+    auto operator=(const TESecureArray& from) noexcept -> auto& {
+        if (this == &from) return *this;
+        MiniCrypt::memcpy(_data, from._data, SIZE);
+        _empty = from._empty;
```
Does it make sense to copy 2 buffers then set `_empty` flag?
Maybe it should be:
```cpp
_empty = from._empty;
if (!_empty)
     MiniCrypt::memcpy(_data, from._data, std::min(SIZE, from.size()));
```

In general, can it be possible that `from.size()` is less than `SIZE`?

DevCheck-StockholmUni commented:
Hi @aregtech, we're researchers from Stockholm University studying communication on GitHub.

A [message](https://github.com/dyfet/areg-sdk/pull/7#discussion_r2484952768) you posted may come across as **rude, disrespectful, or unreasonable** — something that could discourage others from contributing or sharing their ideas:

> can it be possible that the length of the `from` is not equal to `size`? note, that here if the...

You can opt out of messages like this [here](http://otree-extra-url-d0bb646a8437.herokuapp.com/?hash_id=1a4c76774a228424fca1d5b74fcd1525335d7ac7f1a579dbdd95178395cf01fd&opt_out=own&source=github).

**If you are the moderator of this repository, you can opt out of messages like this for the entire repo [here](http://otree-extra-url-d0bb646a8437.herokuapp.com/?hash_id=1a4c76774a228424fca1d5b74fcd1525335d7ac7f1a579dbdd95178395cf01fd&opt_out=all&source=github).**


>[!NOTE]
>Participate in our [four-question survey](http://otree-extra-url-d0bb646a8437.herokuapp.com/?hash_id=1a4c76774a228424fca1d5b74fcd1525335d7ac7f1a579dbdd95178395cf01fd&opt_out=none&source=github) to provide feedback, and [follow us](https://github.com/huentat/GitHub_Research) for project updates!

---

This message was flagged by an automated system and reviewed using an AI tool to check for tone and clarity concerns.

We’re not here to judge. Our goal is to help make collaboration smoother for everyone.

Messages like this can sometimes cause stress or burnout for others involved in the project.

Here are a few quick tips to make messages clearer and easier to engage with:

- Rephrase strong or absolute language
- Explain your reasoning, or suggest an alternative
- Use tools like AI writing assistants to help reword or check tone

Thanks for contributing to GitHub and helping improve collaboration for everyone!

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    template <typename BINARY>
+    auto operator=(const BINARY& from) noexcept -> TESecureArray& {
```
what is the difference between
```cpp
template <typename BINARY>
auto operator=(const BINARY& from) noexcept -> TESecureArray&;
```
and
```cpp
auto operator=(const TESecureArray& from) noexcept -> auto&;
```
?
Does it mean that `BINARY` can be any other type, which has these 2 methods `size()` and `data()`?

dyfet replied:
Indeed, that was the idea behind this kind of adapter. But again its something idiomatic to things I typically do in my codebase...

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+        _erase();
+    }
+
+    auto operator=(const TESecureArray& from) noexcept -> auto& {
```
If here we are sure that `this->SIZE == from.SIZE`, for clarity isn't it better to declare like this?
```cpp
    auto operator=(const TESecureArray<SIZE>& from) noexcept -> auto&;
```

dyfet replied:
I think the code block in question is gone now...

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    auto operator^=(const TESecureArray& other) noexcept -> TESecureArray& {
+        for (unsigned pos = 0; pos < SIZE; ++pos) {
```
do we need to set here hook?
```cpp
if (this != &other)
{
    ......
}
```

same about other operators.

dyfet replied:
I am leaving it for now...

aregtech commented on framework/areg/crypto/TESecureArray.hpp:
```diff
@@ -0,0 +1,177 @@
+            auto val = uint8_t(_data[pos++]);
+            out.push_back(hex[val >> 4]);
+            out.push_back(hex[val & 0x0f]);
+        }
+        return out;
+    }
```
It seems this method is similar to
```cpp
auto NECrypto::EncodeHex(std::string_view input) noexcept -> std::string {
    constexpr char hex[] = "0123456789ABCDEF";
    std::string out;
    out.reserve(input.size() * 2);
    for (const auto& b : input) {
        auto val = uint8_t(b);
        out.push_back(hex[val >> 4]);
        out.push_back(hex[val & 0x0f]);
    }
    return out;
}
```

Does it make sense instead of having duplication of algorithms, have a generic method, so that both,  `NECrypto::EncodeHex()` and `TESecureArray<SIZE>::toHex()`, call same generic method to convert to hex?

dyfet replied:
Since we did add the minicrypt helper, we probably could do that...

dyfet replied:
To hex returns a hex string of the content.  This is kind of what CryptoHelper ToHex is also for....

aregtech reviewed (changes requested):
Please see my additional comments.

dyfet committed:
Clear out unsafe constructor, remove explicit

dyfet committed:
This removes binary type adapter templating

I use this in my code base a lot for adapting different binary
types, but areg does not have this patterm, and there is no
existing code to support.

dyfet committed:
If we rely on copy(), lets make sure its actually correct :)

dyfet committed:
If we rely on copy  argument order should be reversed and defaulted

dyfet committed:
Let's unify hex encoding...

aregtech reviewed (approved):
Thank you very much. Looks nice and compact.
